kubeadm: implement deletion of multiple tokens

bart0sh opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Currently kubeadm doesn't allow to delete more than one bootstrap token at once. This PR implements this feature.

**Does this PR introduce a user-facing change?**:
```release-note
kubeadm: implement deletion of multiple bootstrap tokens at once
```

bart0sh committed:
kubeadm: implement deletion of multiple tokens

bart0sh commented:
/cc @neolit123

bart0sh commented:
/test pull-kubernetes-integration

yagonobre commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -158,11 +158,11 @@ func NewCmdToken(out io.Writer, errW io.Writer) *cobra.Command {
 
 	deleteCmd := &cobra.Command{
-		Use:                   "delete [token-value]",
+		Use:                   "delete token-value [token-value] ...",
```
Maybe `"delete [token-value] ..."`

bart0sh replied:
delete [token value] ... would mean that delete without any parameters is a valid command.
I'll change it to "delete token-value ..." I think. Any objections?

yagonobre commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -296,27 +296,29 @@ func RunListTokens(out io.Writer, errW io.Writer, client clientset.Interface) er
-	if err := client.CoreV1().Secrets(metav1.NamespaceSystem).Delete(tokenSecretName, nil); err != nil {
-		return errors.Wrap(err, "failed to delete bootstrap token")
+		tokenSecretName := bootstraputil.BootstrapTokenSecretName(tokenID)
+		klog.V(1).Infoln("[token] deleting token")
```
Can you print the token id here?

bart0sh replied:
I agree with @rosti here. I'll remove this line.

yagonobre commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -296,27 +296,29 @@ func RunListTokens(out io.Writer, errW io.Writer, client clientset.Interface) er
+// RunDeleteTokens removes a bootstrap tokens from the server.
+func RunDeleteTokens(out io.Writer, client clientset.Interface, tokenIDsOrTokens []string) error {
+	for _, tokenIDOrToken := range tokenIDsOrTokens {
+		// Assume the given first argument is a token id and try to parse it
```
Can you remove or rewrite this comment?

bart0sh replied:
ok, will do. thanks.

yagonobre reviewed (commented):
Thanks @bart0sh 
/priority important-longterm

neolit123 reviewed (commented):
@kubernetes/sig-cluster-lifecycle-pr-reviews 

@bart0sh thanks for the PR. please ping the above team ^ on kubeadm PRs too.

rosti reviewed (commented):
Thanks for the contribution @bart0sh !
I am, however, a bit reserved about this. Having multiple active tokens is not a very common case and deleting more than one at a time is even less common operation.
What if deleting one of the tokens fails for some reason? Do we delete all tokens that we can, do we stop now (potentially deleting only some) or we don't delete anything (if possible)?
Having the user decide on this is way better than us imposing a policy on it.

@bart0sh @fabriziopandini @timothysc @neolit123 WDYT?

bart0sh commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -296,27 +296,29 @@ func RunListTokens(out io.Writer, errW io.Writer, client clientset.Interface) er
+	for _, tokenIDOrToken := range tokenIDsOrTokens {
+		// Assume the given first argument is a token id and try to parse it
+		tokenID := tokenIDOrToken
+		klog.V(1).Infoln("[token] parsing token ID")
```
yep, makes sense. I'll add token id wherever possible.

bart0sh commented:
@rosti thank you for the review. Having a possibility to delete more than one token at a time doesn't make it mandatory. People can continue deleting one token if they wish so. Why not to give them such a possibility?

P.S. I've updated the PR addressing the issues you and @yagonobre pointed to. Please review again.

dixudx commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -158,11 +158,11 @@ func NewCmdToken(out io.Writer, errW io.Writer) *cobra.Command {
 
 	deleteCmd := &cobra.Command{
-		Use:                   "delete [token-value]",
+		Use:                   "delete token-value ...",
```
I'd prefer to the old way. Better not changing the style we're using in current kube command lines.

dixudx replied:
And this usage is misleading. Users may think `token-value` is another new subcommand. But actually it is not.

bart0sh replied:
Well, I'd say that current syntax is misleading. It suggests that token-value is optional as [it's in square brackets](https://en.wikipedia.org/wiki/Usage_message#Pattern).

Anyway, how do you want it to be? ```delete [token-value] ...``` ?

rosti replied:
+1 for @dixudx 's comment. For better or worse, that's our current way of doing things in help. So we have to stick to it for now and, possibly, change things in another PR.

bart0sh replied:
ok, makes sense to me. will chang it to ```delete [token-value] ...```

dixudx reviewed (commented):
Others lgtm.

bart0sh commented:
@rosti you didn't sound negative. Sorry if my answer made you to think so.
Current code exits on any errors, so if even one token can't be deleted kubeadm throws an error and exists. I can change it if reviewers prefer "try to delete all" approach.

bart0sh commented:
@kubernetes/sig-cluster-lifecycle-pr-reviews updated, please review again. Thanks.

neolit123 commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -158,11 +158,11 @@ func NewCmdToken(out io.Writer, errW io.Writer) *cobra.Command {
 		Short:                 "Delete bootstrap tokens on the server.",
 		Long: dedent.Dedent(`
-			This command will delete a given bootstrap token for you.
+			This command will delete a given bootstrap tokens for you.
```
minor typo: `a` with plural `tokens`.

possibly:
`This command will delete a list of bootstrap tokens`

bart0sh replied:
done, thanks.

neolit123 commented:
@rosti @bart0sh 

> We just need to choose a policy of "What happens when one token cannot be deleted?"

seems to me that we are stopping on the first encountered error with the current code.
i'm OK either way.

bart0sh commented:
/test pull-kubernetes-integration

bart0sh commented:
/test pull-kubernetes-integration

bart0sh commented:
@neolit123 @rosti any more objections? if not, please lgtm&approve.

neolit123 reviewed (approved):
/approve

rosti reviewed (approved):
/lgtm
